src/dev/riscv/Plic.py

```diff
@@ -89,12 +96,26 @@ def generateDeviceTree(self, state):
 
         cpus = self.system.unproxy(self).cpu
         int_extended = list()
-        for cpu in cpus:
-            phandle = int_state.phandle(cpu)
-            int_extended.append(phandle)
-            int_extended.append(0xB)
-            int_extended.append(phandle)
-            int_extended.append(0x9)
+        if self.n_contexts != 0:
+            for cpu in cpus:
+                phandle = int_state.phandle(cpu)
+                int_extended.append(phandle)
+                int_extended.append(0xB)
+                int_extended.append(phandle)
+                int_extended.append(0x9)
+        elif self.hart_config != "":
```

Comment:
This needs to be updated in src/python/gem5/components/boards/riscv_board.py and src/python/gem5/prebuilt/riscvmatched/riscvmatched_board.py as well.

The duplicated code is definitely a problem (apologies for asking you to make the updates). Unfortunately, I don't think that having the code for producing the device tree in the SimObject configuration files can work in general.

Reply:
> This needs to be updated in src/python/gem5/components/boards/riscv_board.py and src/python/gem5/prebuilt/riscvmatched/riscvmatched_board.py as well.

Done

> The duplicated code is definitely a problem (apologies for asking you to make the updates). Unfortunately, I don't think that having the code for producing the device tree in the SimObject configuration files can work in general.

Yes, I think it will be use only in the `configs/example/riscv/fs_linux.py`, right?

Reply:
Honestly, I'm not sure how/where this code is used, which is another problem with having code in these files :)